### PR TITLE
feat(get-modified-packages): support retrieving diffs without fetching all

### PR DESCRIPTION
**Pull request by xmfcx**
## Description

Continuation from:
- https://github.com/autowarefoundation/autoware.universe/pull/7408

- Renaming the input parameter name from `base-branch` to `base-ref`. (Mostly it defaults to `main`)
   - I think it is not used anywhere.
   - But we might use it for `merge_group` triggers in the future
- It will run: `git fetch origin ${{ inputs.base-ref }}` to make sure it has common merge-base commit to get diffs
- And now it is possible to only fetch the necessary commits instead of entire history of the repository. (For universe it allows time saving from about 2m30s to 6s)
- Separate the `get-modified-packages` action into 2 steps so it will report if it fails
   - Also exit with failure on the sh script.
   - `set -e` is removed from the script because it is set by default on github actions.

### Failure condition

image

It fails on failure.
https://github.com/autowarefoundation/autoware.universe/actions/runs/9482100138/job/26126284417?pr=7340#step:7:29

### Success condition

image

And it is successful on success:
https://github.com/autowarefoundation/autoware.universe/actions/runs/9482151939/job/26126437257?pr=7340#step:7:35

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by xmfcx** [comment]
I've fixed it.
My solution:

Job:
https://github.com/autowarefoundation/autoware.universe/actions/runs/9470181454/job/26090502061?pr=7340

Workflow: https://github.com/autowarefoundation/autoware.universe/actions/runs/9470181454/workflow?pr=7340#L33-L51

Github action PR:
https://github.com/autowarefoundation/autoware-github-actions/pull/299/files

Step by step:
First part
base-branch in the action is a bad idea because it hides the commit sha of the feature branch's base, it is just "main"
I've replaced it with base-sha, which is github.event.pull_request.base.sha  and it is much more simple to use and fits better with the code (we don't need to add origin/ before it)
It works with existing jobs without problems

Second part
In order to fetch only what we need, we need to do it like this:
    steps:
      - name: "Set PR fetch depth"
        run: echo "PR_FETCH_DEPTH=$(( ${{ github.event.pull_request.commits }} + 1 ))" >> "${GITHUB_ENV}"

      - name: "Checkout PR branch and all PR commits"
        uses: actions/checkout@v4
        with:
          ref: ${{ github.event.pull_request.head.sha }}
          fetch-depth: ${{ env.PR_FETCH_DEPTH }}
That's it, it will fetch only what we need. And everything will work, you can check the job link in this post.

**Comment by xmfcx** [comment]
This solution works for both `fetch-depth: ${{ env.PR_FETCH_DEPTH }}` and for `fetch-depth: 0` so it is safe to merge.

You can test it with existing PRs if you need confirmation using `uses: autowarefoundation/autoware-github-actions/get-modified-packages@v-test-shallow-copy` tag. (Will remove it once this is merged.)

**Comment by xmfcx** [comment]
[Screenshot from the job link above](https://github.com/autowarefoundation/autoware.universe/actions/runs/9470181454/job/26090502061?pr=7340)
image

**Comment by HansRobo** [comment]
Thank you for taking over the job from me and completing it!
BTW, this is a breaking change and will affect people who specify `base-branch` in `get-modified-packages@v1`.
Would you consider bumping the repository major version?

**Comment by xmfcx** [comment]
@HansRobo 

If everyone used it with the default "base-branch" parameter, it won't break existing functionality, anyone using the old version can keep using the new version without modifying any code and it will work the same.

But if someone changed the "base-branch" parameter (which is `github.base_ref` which is `main` by default for most cases), then it won't work since this parameter no longer exists. But I've never seen anyone passing that variable to it. Everyone uses it to get the diffs from the place they've branched out. I think we could even remove that parameter completely, what do you think?

**Comment by xmfcx** [comment]
We can still bump versions but then it will bump the versions of everything else in this repository too. Then any new change on them needs to be updated on the repository sides too (to v2) even if they were not changed. :(

**Comment by xmfcx** [comment]
I've removed the input parameter because I don't see a use for it.
Also updated the readme to have both fetch-all and fetch-enough versions.

**Comment by HansRobo** [comment]
@xmfcx 
Thank you for the discussion!
Considering the pros and cons, it seems better not to bump the major version.

File: get-modified-packages/README.md [hunk 1]
```diff
@@ -4,13 +4,15 @@ This action get the list of ROS packages modified in the pull request.
...
 ```yaml
 jobs:
   get-modified-packages:
     runs-on: ubuntu-latest
     steps:
       - name: Check out repository
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
         with:
           fetch-depth: 0
```
Comment (HansRobo) [inline get-modified-packages/README.md:17]:
```suggestion
          fetch-depth: 1
```

Comment (HansRobo) [reply get-modified-packages/README.md:17]: Oh, I see.
I lived in the days when we did git fetch in script.
Nowadays it is done by tweaking in `actions/checkout`.

**Comment by xmfcx** [comment]
image

Ok now it fails on failure. I will test again to see if it passes on success.

**Comment by xmfcx** [comment]
And it is successful on success:
image

**Review by HansRobo — APPROVED** [review]
Looks good to me.
Thank you for working hard on this task!!

**Comment by xmfcx** [comment]
To confirm, it is working well for regular PRs in the universe too:
- https://github.com/autowarefoundation/autoware.universe/actions/runs/9483164187/job/26129625975#step:7:35

**Comment by HansRobo** [comment]
I have only found one person who uses the `base-branch` argument... and that is me haha :laughing:.
https://github.com/ibis-ssl/crane/blob/8a0fae59b45e578b53064878819aa65cb2af8cd1/.github/workflows/build_test.yaml#L70
Don't worry, I'll fix it according to the new specifications.

**Pull request merged**